src/ucs/datastruct/ptr_map.h

```diff
@@ -0,0 +1,107 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "khash.h"
+#include <ucs/sys/compiler.h>
+#include <ucs/type/status.h>
+
+#include <stdint.h>
+
+#define UCS_PTR_MAP_KEY_INVALID    UINT64_MAX
+#define UCS_PTR_MAP_KEY_ID_FLAG    0x1
+
+
+/* Hash map to find pointer by its key */
+typedef uint64_t ucs_ptr_map_key_t;
+KHASH_TYPE(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*);
+typedef khash_t(ucs_ptr_map_impl) ucs_ptr_hash_t;
+KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+
+typedef struct ucs_ptr_map {
+    uint64_t        cntr;
+    ucs_ptr_hash_t  hash;
+} ucs_ptr_map_t;
+
+
+ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map);
+
+void ucs_ptr_map_destroy(ucs_ptr_map_t *map);
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key = (uintptr_t)ptr;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+    ucs_assert_always(!(key & UCS_PTR_MAP_KEY_ID_FLAG));
+
+    return key;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put_safe(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key;
+    khiter_t iter;
+    int ret;
+
+    do {
+        /* wrap around counter */
+        do {
+            key = ((++map->cntr) << 1) | UCS_PTR_MAP_KEY_ID_FLAG;
+            if (ucs_unlikely(key == UCS_PTR_MAP_KEY_INVALID)) {
+                map->cntr = 0;
+                continue;
+            }
+        } while(0);
```

Comment:
minor: may be use for(;;) loop?
because loop do{ ... continue } while(0) - it is really strange :)

Reply:
IMHO do/while(0) loop looks more linear